File: java/core/src/test/java/dev/varar/core/ParamDiffTest.java
package dev.varar.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Translated from {@code var-core/tests/param-diff.test.ts}. */
class ParamDiffTest {

    private static final String SOURCE = "I should have 3 cukes in my big belly";

    private static Span span(int start, int end) {
        return Span.spanFromOffsets(SOURCE, start, end);
    }

    @Test
    void allElementsEqualEveryCellOk() {
        List<CellDiff> diffs = ParamDiff.compareParams(
                List.of(3, "big"), List.of(3, "big"), List.of(span(14, 15), span(31, 34)), List.of("3", "big"));
        assertTrue(diffs.stream().allMatch(CellDiff::ok));
    }

    @Test
    void oneMismatchingElementThatCellIsNotOkWithExpectedActual() {
        List<CellDiff> diffs = ParamDiff.compareParams(
                List.of(4, "big"), List.of(3, "big"), List.of(span(14, 15), span(31, 34)), List.of("3", "big"));
        assertEquals("arg 1", diffs.get(0).column());
        assertEquals("3", diffs.get(0).expected());
        assertEquals("4", diffs.get(0).actual());
        assertFalse(diffs.get(0).ok());
        assertEquals("arg 2", diffs.get(1).column());
        assertTrue(diffs.get(1).ok());
    }

    @Test
    void objectActualsCompareStructurallyAcrossReferences() {
        List<CellDiff> diffs = ParamDiff.compareParams(
                List.of(Map.of("iso", "NO")), List.of(Map.of("iso", "NO")), List.of(span(0, 2)), List.of("NO"));
        assertTrue(diffs.get(0).ok());
    }
}

File: rust/core/tests/param_diff_test.rs
//! Port of `ParamDiffTest.java` / `param-diff.test.ts`.

mod common;

use common::vmap;
use varar_core::param_diff::compare_params;
use varar_core::span::Span;
use varar_core::value::Value;

const SOURCE: &str = "I should have 3 cukes in my big belly";

fn span(start: usize, end: usize) -> Span {
    Span::from_offsets(SOURCE, start, end)
}

#[test]
fn all_elements_equal_every_cell_ok() {
    let diffs = compare_params(
        &[Value::Int(3), Value::from("big")],
        &[Value::Int(3), Value::from("big")],
        &[span(14, 15), span(31, 34)],
        &["3".to_string(), "big".to_string()],
    );
    assert!(diffs.iter().all(|d| d.ok));
}

#[test]
fn one_mismatching_element_that_cell_is_not_ok_with_expected_actual() {
    let diffs = compare_params(
        &[Value::Int(4), Value::from("big")],
        &[Value::Int(3), Value::from("big")],
        &[span(14, 15), span(31, 34)],
        &["3".to_string(), "big".to_string()],
    );
    assert_eq!("cell 1", diffs[0].column);
    assert_eq!("3", diffs[0].expected);
    assert_eq!("4", diffs[0].actual);
    assert!(!diffs[0].ok);
    assert_eq!("cell 2", diffs[1].column);
    assert!(diffs[1].ok);
}

#[test]
fn object_actuals_compare_structurally_across_references() {
    let diffs = compare_params(
        &[vmap(vec![("iso", Value::from("NO"))])],
        &[vmap(vec![("iso", Value::from("NO"))])],
        &[span(0, 2)],
        &["NO".to_string()],
    );
    assert!(diffs[0].ok);
}
